smartsim/_core/mli/infrastructure/worker/tensorflow_worker.py

```diff
@@ -0,0 +1,316 @@
+    ) -> LoadModelResult:
+        """Given a loaded MachineLearningModel, ensure it is loaded into
+        device memory.
+
+        :param request: The request that triggered the pipeline
+        :param device: The device on which the model must be placed
+        :returns: LoadModelResult wrapping the model loaded for the request
+        :raises ValueError: If model reference object is not found
+        :raises RuntimeError: If loading and evaluating the model failed
+        """
+        if fetch_result.model_bytes:
+            model_bytes = fetch_result.model_bytes
+        elif batch.raw_model and batch.raw_model.data:
+            model_bytes = batch.raw_model.data
+        else:
+            raise ValueError("Unable to load model without reference object")
+
+        try:
+            graph_def = tf.compat.v1.GraphDef()
+            graph_def.ParseFromString(model_bytes)
+
+            # pylint: disable-next=not-context-manager
+            with tf.Graph().as_default() as graph, tf.device(device):
+                tf.import_graph_def(graph_def, name="")
+            ops = graph.get_operations()
+        except Exception as e:
+            raise RuntimeError(
+                "Failed to load and evaluate the model: "
+                f"Model key {batch.model_id.key}, Device {device}"
+            ) from e
+
+        input_layers = set()
+        for operation in ops:
+            if operation.type == "Placeholder":
+                logger.debug(
+                    f"Input op name: {operation.name}, "
+                    f"output shape: {operation.outputs[0].get_shape()}"
+                )
+                input_layers.add(f"{operation.name}:0")
+
+
+        # Code initially taken from
+        # apple.github.io/coremltools/docs-guides/source/tensorflow-1-workflow.html
+        output_tensors = set()
+        input_tensors = set()
+        for operation in ops:
+            for x in operation.inputs:
+                input_tensors.add(x.name)
+        for operation in ops:
+            if len(operation.outputs) > 0:
+                x = operation.outputs[0]
+                potential_names = [x.name]
+                name_split = x.name.split(":")
+                potential_names.append(":".join((name_split[0]+"/resource", name_split[-1])))
+                if all(name not in input_tensors for name in potential_names):
+                    logger.debug(
+                        f"Output tensor name: {x.name}, "
+                        f"tensor shape: {x.get_shape()}, "
+                        f"parent op type: {operation.type}"
+                    )
+                    output_tensors.add(x.name)
+
+        with tf.device(device):
+            result = LoadModelResult(
+                tf.compat.v1.Session(graph=graph),
+                list(input_layers),
+                list(output_tensors),
+            )
+        return result
+
+    @staticmethod
+    def transform_input(
```

Comment:
In general, this method seems super dense and difficult to parse.  Although, I think the logic is correct is there any clear way to add some secondar structs or objects to make this easier to test and grok?

Reply:
Yeah, you're right, and it also shares a lot of code with the Torch version -- although I don't know how much will be shared by other workers. We may create some helper methods - but also possibly introduce a class between TensorFlowWorker and MachineLearningWorkerBase, for those workers which can concatenate tensors this way? Thoughts?